### ISSUE TITLE
[FR] [RFC] add Sequential.append & .extend

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#20117]
A common pattern people use in `__init__` of a `nn.Module` is to build a `list` first and then feed it into `nn.Sequential` because `Sequential` doesn't support many handy methods existing on `list`. E.g., 

```py
def __init__(self, logres):
    super().__init__():
    layers = [
          nn.Conv2d(3, 64, kernel_size=11, stride=4, padding=2),
          nn.ReLU(inplace=True),
          nn.MaxPool2d(kernel_size=3, stride=2),
    ]

    for _ in range(logres):
         layers.extend([
              nn.Conv2d(3, 64, kernel_size=11, stride=4, padding=2),
              nn.ReLU(inplace=True),
              nn.MaxPool2d(kernel_size=3, stride=2),
        ])
    self.layers = nn.Sequential(*layers)
```

This is totally unnecessary if we can just provide `.append` and `.extend` on `Sequential`.

**Comment by crcrpar** [comment]
FYI: [chainer.Sequential](https://docs.chainer.org/en/latest/reference/generated/chainer.Sequential.html#chainer.Sequential) supports those methods and could be a reference.

**Comment by SsnL** [comment]
@crcrpar Thanks for the reference! Chainer also has `insert`, `pop`, and `repeat`, all of which could be useful. I'm curious about `flatten` though, what's the use of that? :)

**Comment by crcrpar** [comment]
TL; DR: the method is for better performance.

In Chainer, `Link`, the base class of `Sequential`, `Chain`, and concrete NN layers w/ learnable parameters such as Conv & BN has `repeat` method that creates a new `Sequential` object from itself [docs](https://docs.chainer.org/en/stable/reference/generated/chainer.Link.html#chainer.Link.repeat).
e.g. `Linear(...).repeat(2)` is equivalent to `Sequential(Linear(...), Linear(...))`.

So, `Sequential(LayerA(), LayerB()).repeat(2)` returns `Sequential(Sequential(LayerA(), LayerB()), Sequential(LayerA(), LayerB())`, not `Sequential(LayerA(), LayerB(), LayerA(), LayerB())`.

But the latter is better from the perspective of overhead and readability and `flatten()` method can take care of this.

**Comment by SsnL** [comment]
I see. Thanks!

**Comment by codexetreme** [comment]
Can I work on this? I would like to implement the features discussed here :)

**Comment by apaszke** [comment]
I think it's reasonable to add that, but I'm not convinced about the `.repeat()` and `.flatten()` API. Let's leave that out.

**Comment by apaszke** [comment]
@codexetreme yes, please go ahead!@codexetreme yes, please go ahead!

**Comment by SsnL** [comment]
Yep I agree with @apaszke . `append`, `extend`, `insert`, and `pop` would be great!

**Comment by codexetreme** [comment]
Cool, ill work on `append`, `extend`, `insert`, `pop`. I'll submit a basic PR then keep pushing work to it and ask you guys if it's going the way it's supposed to ( since this is only my second issue so far, so I will need some input)

**Comment by crcrpar** [comment]
I agree. `repeat` and `flatten` might confuse users.

**Comment by vadimkantorov** [comment]
Also nn.Sequential() + nn.Sequential() would be nice to have to concat two sequentials
